File: python/ccxt/static_dependencies/starknet/hash/poseidon.py
# Ported from ts/src/static_dependencies/scure-starknet/index.ts.

from hashlib import sha256

from ..constants import FIELD_PRIME


RATE = 2
CAPACITY = 1
ROUNDS_FULL = 8
ROUNDS_PARTIAL = 83
MDS_SMALL = [
    [3, 1, 1],
    [1, -1, 1],
    [1, 1, -2],
]


def _field(value):
    return value % FIELD_PRIME


def _round_constant(name, index):
    digest = sha256((name + str(index)).encode()).digest()
    return _field(int.from_bytes(digest, 'big'))


def _round_constants():
    constants = []
    rounds = ROUNDS_FULL + ROUNDS_PARTIAL
    width = RATE + CAPACITY
    for i in range(rounds):
        row = []
        for j in range(width):
            row.append(_round_constant('Hades', width * i + j))
        constants.append(row)
    return constants


ROUND_CONSTANTS = _round_constants()
MDS = [[_field(value) for value in row] for row in MDS_SMALL]


def _sbox(value):
    return pow(value, 3, FIELD_PRIME)


def _poseidon_round(values, is_full, index):
    values = [_field(value + ROUND_CONSTANTS[index][i]) for i, value in enumerate(values)]
    if is_full:
        values = [_sbox(value) for value in values]
    else:
        values[-1] = _sbox(values[-1])
    result = []
    for row in MDS:
        acc = 0
        for i, value in enumerate(values):
            acc += row[i] * value
        result.append(_field(acc))
    return result


def poseidon_hash(values):
    width = RATE + CAPACITY
    if len(values) != width:
        raise ValueError('Poseidon: wrong values length')
    values = [_field(value) for value in values]
    round_index = 0
    half_rounds_full = ROUNDS_FULL // 2
    for _ in range(half_rounds_full):
        values = _poseidon_round(values, True, round_index)
        round_index += 1
    for _ in range(ROUNDS_PARTIAL):
        values = _poseidon_round(values, False, round_index)
        round_index += 1
    for _ in range(half_rounds_full):
        values = _poseidon_round(values, True, round_index)
        round_index += 1
    return values


def poseidon_hash_many(values):
    padded = list(values)
    padded.append(1)
    while len(padded) % RATE != 0:
        padded.append(0)
    state = [0] * (RATE + CAPACITY)
    for i in range(0, len(padded), RATE):
        for j in range(RATE):
            state[j] += padded[i + j]
        state = poseidon_hash(state)
    return state[0]

File: cs/ccxt/static/StarkSharp/StarkSharp.Hash/Poseidon.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Numerics;
using System.Security.Cryptography;
using System.Text;

namespace ccxt;

public static class StarknetPoseidon
{
    private const int Rate = 2;
    private const int Capacity = 1;
    private const int RoundsFull = 8;
    private const int RoundsPartial = 83;
    private static readonly BigInteger FieldPrime = ParseHex("0800000000000011000000000000000000000000000000000000000000000001");
    private static readonly BigInteger[][] RoundConstants = BuildRoundConstants();
    private static readonly BigInteger[][] Mds = new BigInteger[][]
    {
        new BigInteger[] { Field(3), Field(1), Field(1) },
        new BigInteger[] { Field(1), Field(-1), Field(1) },
        new BigInteger[] { Field(1), Field(1), Field(-2) },
    };

    private static BigInteger Field(BigInteger value)
    {
        var result = value % FieldPrime;
        return result.Sign < 0 ? result + FieldPrime : result;
    }

    private static BigInteger ParseHex(string hex)
    {
        hex = hex.StartsWith("0x", StringComparison.OrdinalIgnoreCase) ? hex[2..] : hex;
        return BigInteger.Parse("00" + hex, NumberStyles.AllowHexSpecifier);
    }

    private static BigInteger ToBigInteger(object value)
    {
        if (value is BigInteger bigInteger)
        {
            return bigInteger;
        }
        var valueString = value.ToString();
        if (valueString.StartsWith("0x", StringComparison.OrdinalIgnoreCase))
        {
            return ParseHex(valueString);
        }
        return BigInteger.Parse(valueString, CultureInfo.InvariantCulture);
    }

    private static BigInteger RoundConstant(string name, int index)
    {
        using var sha256 = SHA256.Create();
        var digest = sha256.ComputeHash(Encoding.UTF8.GetBytes(name + index.ToString(CultureInfo.InvariantCulture)));
        var hex = BitConverter.ToString(digest).Replace("-", "").ToLowerInvariant();
        return Field(ParseHex(hex));
    }

    private static BigInteger[][] BuildRoundConstants()
    {
        var rounds = RoundsFull + RoundsPartial;
        var width = Rate + Capacity;
        var constants = new BigInteger[rounds][];
        for (var i = 0; i < rounds; i++)
        {
            constants[i] = new BigInteger[width];
            for (var j = 0; j < width; j++)
            {
                constants[i][j] = RoundConstant("Hades", width * i + j);
            }
        }
        return constants;
    }

    private static BigInteger Sbox(BigInteger value)
    {
        return BigInteger.ModPow(value, 3, FieldPrime);
    }

    private static BigInteger[] PoseidonRound(BigInteger[] values, bool isFull, int index)
    {
        values = values.Select((value, i) => Field(value + RoundConstants[index][i])).ToArray();
        if (isFull)
        {
            values = values.Select(Sbox).ToArray();
        }
        else
        {
            values[^1] = Sbox(values[^1]);
        }
        var result = new BigInteger[Mds.Length];
        for (var row = 0; row < Mds.Length; row++)
        {
            var acc = BigInteger.Zero;
            for (var i = 0; i < values.Length; i++)
            {
                acc += Mds[row][i] * values[i];
            }
            result[row] = Field(acc);
        }
        return result;
    }

    private static BigInteger[] PoseidonHash(BigInteger[] values)
    {
        var width = Rate + Capacity;
        if (values.Length != width)
        {
            throw new ArgumentException("Poseidon: wrong values length");
        }
        values = values.Select(Field).ToArray();
        var roundIndex = 0;
        var halfRoundsFull = RoundsFull / 2;
        for (var i = 0; i < halfRoundsFull; i++)
        {
            values = PoseidonRound(values, true, roundIndex++);
        }
        for (var i = 0; i < RoundsPartial; i++)
        {
            values = PoseidonRound(values, false, roundIndex++);
        }
        for (var i = 0; i < halfRoundsFull; i++)
        {
            values = PoseidonRound(values, true, roundIndex++);
        }
        return values;
    }

    public static BigInteger HashMany(object values)
    {
        var padded = new List<BigInteger>();
        foreach (var value in (IEnumerable)values)
        {
            padded.Add(ToBigInteger(value));
        }
        padded.Add(BigInteger.One);
        while ((padded.Count % Rate) != 0)
        {
            padded.Add(BigInteger.Zero);
        }
        var state = new BigInteger[Rate + Capacity];
        for (var i = 0; i < padded.Count; i += Rate)
        {
            for (var j = 0; j < Rate; j++)
            {
                state[j] += padded[i + j];
            }
            state = PoseidonHash(state);
        }
        return state[0];
    }
}
